### PR TITLE
Build misc examples

### LINKED ISSUE #6866: examples/misc/longshort.py is outdated
As of master / numpy 1.11:
```
$ python2 examples/misc/longshort.py
Traceback (most recent call last):
  File "examples/misc/longshort.py", line 24, in <module>
    r1 = mlab.rec_append_fields(r1, 'gains', gains1)
  File "/home/antony/src/extern/matplotlib/lib/matplotlib/mlab.py", line 2389, in rec_append_fields
    newdtype = np.dtype(rec.dtype.descr + list(zip(names, dtypes)))
TypeError: data type not understood

$ python3 examples/misc/longshort.py
Traceback (most recent call last):
  File "examples/misc/longshort.py", line 12, in <module>
    u1 = urllib.urlretrieve('http://ichart.finance.yahoo.com/table.csv?s=AAPL&d=9&e=14&f=2008&g=d&a=8&b=7&c=1984&ignore=.csv')
AttributeError: module 'urllib' has no attribute 'urlretrieve'
```

Basically `mlab.rec_append_fields` has some unicode/bytes issues on Py2 and the imports need to be renamed on Py3 (I guess).

### PR DESCRIPTION
**Pull request by jenshnielsen**
Fixes #6866

urllib has moved from python 2 to 3. file is no more.

In mlab:

iterkeys was being called on the keys instead of the dict.
Unicode strings need to be encoded on python2 to be used in np.dtype

Should probably still add some tests to test_mlab.

### CODE REVIEW

File: lib/matplotlib/mlab.py [hunk 1]
```diff
@@ -2387,8 +2387,10 @@ def rec_append_fields(rec, names, arrs, dtypes=None):
             dtypes = dtypes * len(arrs)
         else:
             raise ValueError("dtypes must be None, a single dtype or a list")
-
-    newdtype = np.dtype(rec.dtype.descr + list(zip(names, dtypes)))
+    old_dtypes = rec.dtype.descr
+    if six.PY2:
+        old_dtypes = [(a[0].encode('utf-8'), a[1]) for a in old_dtypes]
```
Comment (Kojoley) [inline lib/matplotlib/mlab.py]: A bit less cryptic variant: `old_dtypes = [(name.encode('utf-8'), dt) for name, dt in old_dtypes]`

Comment (jenshnielsen) [inline lib/matplotlib/mlab.py]: Agreed but I am not sure this is the right solution anyway. The recarray can handle unicode keys it's only that np.dtype cannot. We should probably do this in a completely different way

**Comment by tacaswell** [comment]
I am inclined to merge this as-is.  It improves things and we should not let the perfect be the enemy of the good here.

**Comment by jenshnielsen** [comment]
I will at least implement @Kojoley's suggestion

**Comment by jenshnielsen** [comment]
Implemented @Kojoley's sugestion. The proper solution is most likely to use `fromrecords` to build the recarrays but I don't want to make any more modifications to this code which should probably be deprecated and removed instead

**Pull request merged**